File: finetune.py
import os
from dotenv import load_dotenv
import streamlit as st
from langchain_community.document_loaders import PyPDFLoader
from langchain_openai import OpenAIEmbeddings
from langchain_community.vectorstores import FAISS
from langchain_community.chat_models import ChatOpenAI
from langchain.chains import LLMChain
from langchain_core.messages import HumanMessage, AIMessage
from langchain_core.prompts import MessagesPlaceholder
from langchain_core.prompts import ChatPromptTemplate

load_dotenv()

# Constants
DATA_PATH = "./static/Pedagogy_Portfolio.pdf"
API_KEY = os.getenv("OPENAI_API_KEY")

# Load document and create vector store
@st.cache_data
def load_document():
    document_loader = PyPDFLoader(DATA_PATH)
    document = document_loader.load()
    return document[0].page_content  # Extract the full text if it's only one document

@st.cache_data
def create_vector_store(document_text, API_KEY):
    embeddings = OpenAIEmbeddings(openai_api_key=API_KEY)
    vector_store = FAISS.from_texts([document_text], embedding=embeddings)
    vector_store.save_local("faiss_index")
    return vector_store

def load_vector_store():
    embeddings = OpenAIEmbeddings(openai_api_key=API_KEY)
    if os.path.exists("faiss_index"):
        return FAISS.load_local("faiss_index", embeddings=embeddings, allow_dangerous_deserialization=True)
    else:
        document_text = load_document()
        return create_vector_store(document_text, API_KEY)

# Function to set up conversation chain
def get_conversational_chain(API_KEY):
    model = ChatOpenAI(model_name="gpt-3.5-turbo", temperature=0.3, openai_api_key=API_KEY)
    
    prompt = ChatPromptTemplate.from_messages([
        ("system", "You are a friendly and professional customer service assistant for Pedagogy. "
                   "Answer the user's questions based on the provided document information."),
        MessagesPlaceholder(variable_name="chat_history"),
        ("human", "{question}")
    ])
    
    llm_chain = LLMChain(llm=model, prompt=prompt)
    return llm_chain

# User input processing with chat history
def user_input(user_question, API_KEY, vector_store, chat_history, document_text):
    if vector_store is None:
        st.error("Error: Vector store is not loaded correctly.")
        return "Error: Vector store is not loaded correctly."
    
    # Directly use the loaded document content as context
    chain = get_conversational_chain(API_KEY)
    response = chain({"context": document_text, "question": user_question, "chat_history": chat_history})
    
    return response["text"]

# Load the vector store and document text
vector_store = load_vector_store()
document_text = load_document()  # Load the full document content once

# Streamlit App
def main():
    st.title("Pedagogy Q&A Assistant")
    
    # Initialize chat history in session state
    if "chat_history" not in st.session_state:
        st.session_state.chat_history = []

    # User Question
    user_question = st.text_input("Ask a question about Pedagogy:")
    if user_question:
        # Get the response based on user question and chat history
        response = user_input(user_question, API_KEY, vector_store, st.session_state.chat_history, document_text)
        
        # Update the chat history in session state
        st.session_state.chat_history.append(HumanMessage(content=user_question))
        st.session_state.chat_history.append(AIMessage(content=response))
        
        # Display the response
        st.write("**Response:**")
        st.write(response)

# Run the app
if __name__ == "__main__":
    main()

File: finetune.js
import "dotenv/config";
import { ChatOpenAI } from "@langchain/openai";
import { PDFLoader } from "@langchain/document_loaders/fs/pdf";
import { RecursiveCharacterTextSplitter } from "@langchain/textsplitters";
import { MemoryVectorStore } from "@langchain/vectorstores/memory";
import { OpenAIEmbeddings } from "@langchain/openai";
import { createRetrievalChain } from "@langchain/chains/retrieval";
import { ChatPromptTemplate } from "@langchain/prompts";

// Load environment variables
const API_KEY = process.env.OPENAI_API_KEY;
const DATA_PATH = "./static/Pedagogy_Portfolio.pdf";

// Step 1: Load the PDF Document
const loadDocuments = async () => {
  const loader = new PDFLoader(DATA_PATH);
  const docs = await loader.load();
  console.log(`Loaded ${docs.length} pages from PDF.`);
  return docs;
};

// Step 2: Split the Documents
const splitDocuments = async (docs) => {
  const textSplitter = new RecursiveCharacterTextSplitter({
    chunkSize: 300,
    chunkOverlap: 30,
  });
  const textChunks = await textSplitter.splitDocuments(docs);
  console.log(`Split document into ${textChunks.length} chunks.`);
  return textChunks;
};

// Step 3: Create the Vector Store with Embeddings
const createVectorStore = async (textChunks) => {
  const embeddings = new OpenAIEmbeddings({ apiKey: API_KEY });
  const vectorStore = await MemoryVectorStore.fromDocuments(textChunks, embeddings);
  console.log("Vector store created with embeddings.");
  return vectorStore;
};

// Step 4: Set up the RAG Chain and Prompt Template
const getConversationalChain = async (retriever) => {
  const promptTemplate = new ChatPromptTemplate([
    ["system", `
      You are a professional assistant for Pedagogy, an educational consulting firm. Respond to user questions based on the given context.
      Only use information in the provided context to answer. If information is not available, ask the user to reach out to the office.
      Keep responses concise and friendly.
      \n\nContext:\n {context}\n\nQuestion:\n{input}\nAnswer:\n
    `],
    ["user", "{input}"],
  ]);

  const model = new ChatOpenAI({ model: "gpt-3.5-turbo", apiKey: API_KEY });
  const chain = await createRetrievalChain({
    retriever,
    prompt: promptTemplate,
    llm: model,
  });
  return chain;
};

// Step 5: Set up the Question-Answering System
const questionAnswerSystem = async (userQuestion, vectorStore) => {
  const retriever = vectorStore.asRetriever();
  const chain = await getConversationalChain(retriever);

  // Run the chain with user input
  const response = await chain.invoke({ input: userQuestion });
  return response.text || "No relevant answer found in the context.";
};

// Main function
const main = async () => {
  // Load and process documents
  const documents = await loadDocuments();
  const textChunks = await splitDocuments(documents);
  const vectorStore = await createVectorStore(textChunks);

  // Example usage
  const userQuestion = "What services does Pedagogy offer?";
  const answer = await questionAnswerSystem(userQuestion, vectorStore);
  console.log("Answer:", answer);
};

main().catch(console.error);
